src/constbuffer_array_splitter.c

```diff
@@ -208,3 +208,353 @@ CONSTBUFFER_ARRAY_HANDLE constbuffer_array_splitter_split(CONSTBUFFER_ARRAY_HAND
 
     return result;
 }
+
+TARRAY(CONSTBUFFER_ARRAY_HANDLE) constbuffer_array_splitter_split_to_array_of_array(CONSTBUFFER_ARRAY_HANDLE buffers, uint32_t max_buffer_size)
+{
+    TARRAY(CONSTBUFFER_ARRAY_HANDLE) result = NULL;
+
+    if (
+        /* Codes_SRS_CONSTBUFFER_ARRAY_SPLITTER_07_001: [ If buffers is NULL then constbuffer_array_splitter_split_to_array_of_array shall fail and return NULL. ]*/
+        buffers == NULL ||
+        /* Codes_SRS_CONSTBUFFER_ARRAY_SPLITTER_07_002: [ If max_buffer_size is 0 then constbuffer_array_splitter_split_to_array_of_array shall fail and return NULL. ]*/
+        max_buffer_size == 0)
+    {
+        LogError("Invalid args : CONSTBUFFER_ARRAY_HANDLE buffers = %p, size_t max_buffer_size = %" PRIu32,
+            buffers, max_buffer_size);
+    }
+    else
+    {
+        /* Codes_SRS_CONSTBUFFER_ARRAY_SPLITTER_07_003: [ constbuffer_array_splitter_split_to_array_of_array shall create a TARRAY(CONSTBUFFER_ARRAY_HANDLE) by calling TARRAY_CREATE. ]*/
+        TARRAY(CONSTBUFFER_ARRAY_HANDLE) temp = TARRAY_CREATE(CONSTBUFFER_ARRAY_HANDLE)();
+        if (temp == NULL)
+        {
+            /* Codes_SRS_CONSTBUFFER_ARRAY_SPLITTER_07_004: [ If there are any other failures then constbuffer_array_splitter_split_to_array_of_array shall fail and return NULL. ]*/
+            LogError("TARRAY_CREATE(CONSTBUFFER_ARRAY_HANDLE) failed");
+        }
+        else
+        {
+            uint32_t buffer_array_count = 0; //used to store total number of constbuffer_array in result
+            uint32_t buffer_count;
+
+            /* Codes_SRS_CONSTBUFFER_ARRAY_SPLITTER_07_005: [ constbuffer_array_splitter_split_to_array_of_array shall call constbuffer_array_get_buffer_count to get the total number of buffers. ]*/
+            (void)constbuffer_array_get_buffer_count(buffers, &buffer_count);
+
+            /* Codes_SRS_CONSTBUFFER_ARRAY_SPLITTER_07_006: [ constbuffer_array_splitter_split_to_array_of_array shall call TARRAY_ENSURE_CAPACITY to be able to add a CONSTBUFFER_ARRAY_HANDLE entry in the TARRAY(CONSTBUFFER_ARRAY_HANDLE). ]*/
+            if (TARRAY_ENSURE_CAPACITY(CONSTBUFFER_ARRAY_HANDLE)(temp, 1) != 0)
```

Comment:
>   if (TARRAY_ENSURE_CAPACITY(CONSTBUFFER_ARRAY_HANDLE)(temp, 1) != 0) [](http://example.com/codeflow?start=9&length=70)

This will do a realloc behind. It is not necessary. If buffer_count == 0, we can directly create the TARRAY with size 1.

Also for the case where there is no splitting, but we use the entire original array as is we know we have to create it with 1 as capacity.

As a matter of fact we can easily compute in one pass over all the buffers in the original array how many entries we need in the TARRAY so we don't do extra mallocs.

Reply:
changed to one malloc